src/main/kotlin/org/rust/lang/core/CFGBuilder.kt

```diff
@@ -0,0 +1,366 @@
+
+
+    private var result: CFGIndex? = null
+    private val preds: Deque<CFGIndex> = ArrayDeque<CFGIndex>()
+    private val pred: CFGIndex get() = preds.peek()
+    private val loopScopes: Deque<LoopScope> = ArrayDeque<LoopScope>()
+    private val breakableBlockScopes: Deque<BlockScope> = ArrayDeque<BlockScope>()
+
+    private inline fun finishWith(callable: () -> CFGIndex) {
+        result = callable()
+    }
+
+    private fun finishWith(value: CFGIndex) {
+        result = value
+    }
+
+    private fun withLoopScope(loopScope: LoopScope, callable: () -> Unit) {
+        loopScopes.push(loopScope)
+        callable()
+        loopScopes.pop()
+    }
+
+    private fun addAstNode(element: RsElement, preds: List<CFGIndex>): CFGIndex = addNode(CFGNodeData.AST(element), preds)
+
+    private fun addDummyNode(preds: List<CFGIndex>): CFGIndex = addNode(CFGNodeData.Dummy, preds)
+
+    private fun addUnreachableNode(): CFGIndex = addNode(CFGNodeData.Unreachable, emptyList())
+
+    private fun addNode(data: CFGNodeData, preds: List<CFGIndex>): CFGIndex {
+        val newNode = graph.addNode(data)
+        preds.forEach { addContainedEdge(it, newNode) }
+        return newNode
+    }
+
+    fun addContainedEdge(source: CFGIndex, target: CFGIndex) {
+        val data = CFGEdgeData(mutableListOf())
+        graph.addEdge(source, target, data)
+    }
+
+    private fun addReturningEdge(fromIndex: CFGIndex) {
+        val data = CFGEdgeData(loopScopes.map { it.loop }.toMutableList())
+        graph.addEdge(fromIndex, exit, data)
+    }
+
+    private fun straightLine(expr: RsExpr, pred: CFGIndex, subExprs: List<RsExpr?>): CFGIndex {
+        val subExprsExit = subExprs.fold(pred) { pred, subExpr -> process(subExpr, pred) }
+        return addAstNode(expr, listOf(subExprsExit))
+    }
+
+    // This function will be used during break/continue expressions processing
+    // TODO: change return type to Pair<Region.Scope, NodeIndex>
+    private fun findScopeEdge(expr: RsExpr, destination: Destination, kind: ScopeControlFlowKind): CFGIndex? {
+        for (b in breakableBlockScopes) {
+            if (b.block == destination.target) {
+                return when (kind) {
+                    ScopeControlFlowKind.BREAK -> b.breakIndex
+                    ScopeControlFlowKind.CONTINUE -> null
+                }
+            }
+        }
+        for (l in loopScopes) {
+            if (l.loop == destination.target) {
+                return when (kind) {
+                    ScopeControlFlowKind.BREAK -> l.breakIndex
+                    ScopeControlFlowKind.CONTINUE -> l.continueIndex
+                }
+            }
+        }
+
+        return null
+    }
+
+    fun process(element: RsElement?, pred: CFGIndex): CFGIndex {
+        if (element == null) return pred
+
+        val oldPredsSize = preds.size
+        preds.push(pred)
+        element.accept(this)
+        preds.pop()
+        assert(preds.size == oldPredsSize)
+
+        return result!!
```

Comment:
Maybe It's better to use `checkNotNull`:
```kotlin
return checkNotNull(result) { "Optional (lazy) error message" }
```
This method throws `IllegalStateException` instead of `NullPointerException`.